src/python/tests/test_local_functions.py

```diff
@@ -0,0 +1,258 @@
+              f"source code is {_meta_data['format_version']}."
+
+    def tearDown(self):
+        """
+        Write json test file if we are generating resuls.
+        """
+        if generate_local_function_results:
+            # add the format version to the output
+            self.test_sets.insert(0, _meta_data)
+            with open(self.test_file, mode="w") as f:
+                f.write(json.dumps(self.test_sets, indent=4))
+
+    def generate_parameter_sets(self):
+        """
+        Generate all permutations of parameter values except exclusions.
+        """
+
+        par_sets = []
+
+        pars = parameter_grid.keys()
+        for perm in itertools.product(*parameter_grid.values()):
+            my_dict = dict(zip(pars, perm))
+
+            exclude = False
+            for my_exclude in exclude_sets:
+                matches = sum([my_exclude[par] == my_dict[par]
+                               for par in my_exclude])
+                if matches == len(my_exclude):
+                    exclude = True
+                    break
+
+            if exclude:
+                continue
+
+            par_sets.append(my_dict)
+
+        return par_sets
+
+    def generate_base_inputs(self):
+        """
+        Generate sets of sensible random inputs using random_inputs dict.
+
+        This generates just the minimal inputs, out of which a
+        full set of input fields can be generated with a given
+        set of runtime parameters.
+        """
+
+        values = {}
+        rng = np.random.default_rng(self.seed)
+        for field, config in random_inputs.items():
+            vals = (config["max"] - config["min"]) * \
+              rng.random(size=self.n_input_sets) + config["min"]
+            if config["log"]:
+                vals = 10**vals
+            values[field] = vals
+
+        base_inputs = [{field: values[field][i] for field in values}
+                       for i in range(self.n_input_sets)]
+        return base_inputs
+
+    def test_local_functions(self):
+        for test_set in self.test_sets:
+
+            par_set = test_set["parameters"]
+            base_units = test_set["units"]
+
+            my_chemistry = chemistry_data()
+            for par, val in par_set.items():
+                setattr(my_chemistry, par, val)
+
+            # add path to data file here so we only have to save the base name
+            if "grackle_data_file" in par_set:
+                setattr(my_chemistry, "grackle_data_file",
+                        os.path.join(grackle_data_dir, par_set["grackle_data_file"]))
+
+            if generate_local_function_results:
+                my_tests = []
+                base_inputs = self.generate_base_inputs()
+                n_input_sets = len(base_inputs)
+            else:
+                my_tests = test_set["tests"]
+                n_input_sets = len(my_tests)
+            assert n_input_sets == self.n_input_sets
+
+            for itest in range(n_input_sets):
```

Comment:
If I'm understanding this right, this test is structured so that we determine the inputs for all tests in python code, right? In effect, the JSON files are just a "meta-data-rich" way to specify the test inputs, right? (i.e. the inputs aren't actually used by the test).

I just want to clarify because I thought you had talked about doing something different a while back, where you could create a new test by creating a new json file. (It's totally fine if you are not doing -- I just want to make sure I understand).

Reply:
Not exactly. When `GENERATE_PYGRACKLE_TEST_RESULTS=1`, then the inputs are generated by Python code and stored in the json. However, when `GENERATE_PYGRACKLE_TEST_RESULTS=0`, the inputs are read in from the json file directly. This is happening in `LocalFunctionsTest.setUp`. I'll leave a comment pointing it out. Let me know if this isn't what you were referring to.